### ISSUE TITLE
About the use of gather to compute metrics

### ISSUE THREAD

**Issue by jbschiratti** [huggingface/accelerate#226]
First of all, thank you for developing Accelerate! I'm new to it but I already love it, it's a great framework.

I used the code below to train a naive model on MNIST data using 3 GPUs (on a single node/machine). This code uses an `Accuracy` class to compute the epoch-wise accuracy from predictions and labels. It stores predictions/labels at each step to eventually compute accuracy (or another metric, such as ROC AUC) _once_ at the end of an epoch. At each step, `accelerator.gather()` is used to gather all predictions/labels from GPU devices. I added a `print` statement in the `Accuracy` class to check the number of samples used to compute the accuracy. The test set of MNIST is composed of 10 000 samples but the print statement in this class shows that 10 167 samples are actually used. 

Full code: 
```python
from __future__ import print_function

import argparse
import os
import os.path
import threading
from functools import partial

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

from accelerate import Accelerator
from sklearn.metrics import accuracy_score
from torch.optim.lr_scheduler import StepLR
from torch.utils.data import DataLoader
from torchvision import datasets, transforms
from tqdm.auto import tqdm as original_tqdm


class Accuracy:
    """Accuracy score."""
    def __init__(self):
        super().__init__()
        self.__build()

    def __build(self):
        self._lock = threading.Lock()
        self._predictions = []
        self._targets = []

    def reset(self):
        self._predictions.clear()
        self._targets.clear()

    def update(self, output):
        y_pred, y_true = output
        with self._lock:
            self._predictions.append(y_pred)
            self._targets.append(y_true)

    def compute(self):
        with self._lock:
            predictions = torch.cat(self._predictions, dim=0).numpy()
            targets = torch.cat(self._targets, dim=0).numpy()
            print(f'Shapes: predictions {predictions.shape}, targets {targets.shape}')
            return accuracy_score(y_true=targets, y_pred=predictions)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def main():

    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--per_device_eval_batch_size',
                        type=int,
                        default=64,
                        metavar='N',
                        help='The per-device batch size to use for evaluation.')
    parser.add_argument('--per_device_train_batch_size',
                        type=int,
                        default=64,
                        metavar='N',
                        help='The per-device batch size to use for training.')
    parser.add_argument('--epochs',
                        type=int,
                        default=5,
                        metavar='N',
                        help='number of epochs to train (default: 14)')
    parser.add_argument('--lr',
                        type=float,
                        default=1.0,
                        metavar='LR',
                        help='learning rate (default: 1.0)')
    parser.add_argument('--gamma',
                        type=float,
                        default=0.7,
                        metavar='M',
                        help='Learning rate step gamma (default: 0.7)')
    parser.add_argument('--no-cuda',
                        action='store_true',
                        default=False,
                        help='disables CUDA training')
    parser.add_argument('--seed',
                        type=int,
                        default=1,
                        metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval',
                        type=int,
                        default=10,
                        metavar='N',
                        help='how many batches to wait before logging training status')
    parser.add_argument('--out_dir',
                        type=str,
                        help='Path where the trained model will be saved (if not None).')
    args = parser.parse_args()

    torch.manual_seed(args.seed)
    accelerator = Accelerator()
    _is_local_main_process = accelerator.is_local_main_process
    tqdm = partial(original_tqdm, disable=not _is_local_main_process, position=0)

    use_cuda = not args.no_cuda and torch.cuda.is_available()
    kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}

    # TRAIN AND TEST DATASETS/DATALOADERS
    train_transforms = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,)),
    ])

    test_transforms = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,)),
    ])

    with accelerator.main_process_first():
        # We only want to download MNIST data on rank-0

        train_dataset = datasets.MNIST(os.environ['DSDIR'],
                                       train=True,
                                       download=True,
                                       transform=train_transforms)
        print(f'Length of training dataset: {len(train_dataset)}')

        test_dataset = datasets.MNIST(os.environ['DSDIR'],
                                      download=True,
                                      train=False,
                                      transform=test_transforms)
        print(f'Length of test dataset: {len(test_dataset)}')

    train_dataloader = DataLoader(dataset=train_dataset,
                                  batch_size=args.per_device_train_batch_size,
                                  shuffle=True,
                                  **kwargs)

    test_dataloader = DataLoader(dataset=test_dataset,
                                 batch_size=args.per_device_eval_batch_size,
                                 shuffle=True,
                                 **kwargs)

    model = Net().to(accelerator.device)
    optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
    scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
    model, optimizer, train_dataloader, test_dataloader = accelerator.prepare(model, optimizer, train_dataloader, test_dataloader)

    def evaluate(_model, _device, _test_loader, _epoch):
        _model.eval()
        test_losses = []
        test_accuracy = Accuracy()
        example_images = []
        with torch.no_grad():
            for data, target in tqdm(_test_loader, desc=f'eval (epoch {_epoch:03d})'):
                data, target = data.to(_device), target.to(_device)
                output = _model(data)
                loss = F.nll_loss(output, target, reduction='sum')
                test_losses.append(accelerator.gather(loss))
                preds = output.argmax(dim=1, keepdim=True)
                test_accuracy.update((accelerator.gather(preds).detach().cpu(),
                                      accelerator.gather(target).detach().cpu()))

        test_loss = torch.sum(torch.cat(test_losses)) / len(_test_loader.dataset)
        test_acc = test_accuracy.compute()
        test_accuracy.reset()
        return test_acc

    def train_one_epoch(_args, _model, _device, _train_loader, _optimizer, _epoch):
        _model.train()
        for step, batch in enumerate(tqdm(_train_loader, desc=f'train (epoch {_epoch:03d})')):
            data, target = batch
            data, target = data.to(_device), target.to(_device)
            _optimizer.zero_grad()
            output = _model(data)
            loss = F.nll_loss(output, target)
            accelerator.backward(loss)
            _optimizer.step()

    # TRAINING
    for epoch in range(1, args.epochs + 1):
        train_one_epoch(args, model, accelerator.device, train_dataloader, optimizer, epoch)
        eval_accuracy = evaluate(model, accelerator.device, test_dataloader, epoch)
        if _is_local_main_process:
            print(f'Epoch {epoch:02d} / Eval accuracy = {eval_accuracy}')
        scheduler.step()

    # SAVE TRAINED MODEL (OPTIONAL)
    if _is_local_main_process and args.out_dir is not None:
        accelerator.wait_for_everyone()
        unwrapped_model = accelerator.unwrap_model(model)
        unwrapped_model.save(args.out_dir, save_function=accelerator.save)


if __name__ == '__main__':

    main()
```

Print statement after the 1st epoch: 
```
Shapes: predictions (10176, 1), targets (10176,)
```

Why is there a mismatch with the expected number of samples (10 000) and the actual number of samples (10 176)?

**Comment by sgugger** [comment]
Hi there! Thanks for the praise, I'm glad you like the library. Regarding your question, this is because Accelerate will add samples to make sure each process gets the same batch size (otherwise you will get weird errors) by looping back to the beginning of the dataset. So for evaluation, you need to truncate your predictions to the length of the dataset.

We will add more tutorials to explain those caveats in the coming weeks :-)

**Comment by jbschiratti** [comment]
Thank you @sgugger for the quick answer. Indeed, more tutorials on this would be welcome :-) 

Also (on another topic), construction of the dataset is done inside the context defined by `with accelerator.main_process_first():`. Is it good practice to avoid preparing/downloading data on all processes?

**Comment by sgugger** [comment]
Yes, especially since you could get race conditions with various processes trying to write to the same file, and a big slowdown if you try to prepare the data num_processes times.

**Comment by jbschiratti** [comment]
Thank you for your answers @sgugger!
As you mention adding more examples in the coming weeks, could this MNIST example be transformed into an official computer vision example?

**Comment by sgugger** [comment]
Sure thing! Would you like to make a PR with it?

**Comment by jbschiratti** [comment]
Yes, I will start one. Thank you 😃

**Comment by tanyaroosta** [comment]
Hi @sgugger . It would be great if the tutorials for Accelerate can be expanded.  I have used the xxx_no_trainer.py examples (specifically MLM and summarization) so I can have control over the training loop.  I am trying to implement early stop, and my question is around whether I need the "gather" function or not.  Below is my code and my main confusion is if I need to replace with something that involves the gather function, since I noticed an example in the MLM code (accelerator.gather(loss.repeat(args.per_device_eval_batch_size)))

for epoch in range(num_train_epochs):
        model.train()
        for step, batch in enumerate(train_dataloader):
            # training loop

            # update running values
            running_loss += loss.item()
            global_step += 1

            # evaluation step
            if global_step % eval_every == 0:
                model.eval()

                # validation loop
                for step, inputs in enumerate(eval_dataloader):
                    batch.to(accelerator.device)
                    with torch.no_grad():  
                        outputs = model(**batch)
                    
                    loss = outputs.loss                           ----------> Does this need to include gather function?
                   
                    valid_running_loss += loss

                # evaluation
                average_train_loss = running_loss / eval_every
                average_valid_loss = valid_running_loss / len(eval_dataloader)
                train_loss_list.append(average_train_loss)
                valid_loss_list.append(average_valid_loss)
                global_steps_list.append(global_step)

                # resetting running values
                running_loss = 0.0                
                valid_running_loss = 0.0
                model.train()
                
        if early_stopping.step(average_valid_loss):
            break  # early stop criterion is met, we can stop now

**Comment by sgugger** [comment]
If you want the loss to be averaged across all processes, yes you will need to gather them all.

**Comment by tanyaroosta** [comment]
It seems for early stop, that is what I need to do.  If I understand accelerate correctly, it is spreading the training load across available GPUs (in the multi-gpu setting).

**Comment by tanyaroosta** [comment]
@sgugger qq, is/will there be a tutorial that explains a bit more about the gather and unwrap. My main question is when you have multiple gpus and you are training, at what stages does it make sense to call gather, or unwrap.  Reading through the existing material, I couldn't get a clear answer. 

Thanks!

**Comment by jbschiratti** [comment]
Interesting @miquelmarti!

From my understanding, it may be **safer** to instantiate a DataLoader with `drop_last=True`.
In that case, no additional samples are added. Indeed, in [distributed.py](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py) (torch.utils.data), one can read:

```python
        if not self.drop_last:
            # add extra samples to make it evenly divisible
            padding_size = self.total_size - len(indices)
            if padding_size <= len(indices):
                indices += indices[:padding_size]
            else:
                indices += (indices * math.ceil(padding_size / len(indices)))[:padding_size]
        else:
            # remove tail of data to make it evenly divisible.
            indices = indices[:self.total_size]
```

When a DataLoader is instantiated with `drop_last=False`, additional samples will be added.
This issue also affects Accelerate (see above, I noted that metrics such as accuracy can be computed using more samples than the length of the (eval) dataset). The question which remains unclear to me is: how to deal with these additional samples? Following the above discussion, I initiated a PR (#252) to add a computer vision example (training a simple convolutional model on MNIST data). To write this example, I followed @sgugger's suggestion and just truncate model predictions before computing the accuracy (see [here!](https://github.com/huggingface/accelerate/blob/418a6c456356321a3f6daec9d7c0c27e2526e383/examples/mnist_example.py#L218)). Still, I'm not sure this is the right way to do it... If the DataLoader is instantiated with `shuffle=True`, the added samples may not be always at the beginning (or at the end).

**Comment by sgugger** [comment]
> it may be safer to instantiate a DataLoader with drop_last=True.
In this case you are not evaluating on your whole dataset, so the reported metric is also wrong.

> If the DataLoader is instantiated with shuffle=True, the added samples may not be always at the beginning (or at the end).

No, there will be at the end. Accelerate will use the same permutation and loop back at the beginning, so you will still have your whole predictions when truncating to the dataset length, just in a random order.

**Comment by ashutoshsaboo** [comment]
Hey @sgugger , first of all huge thanks for all your efforts in this library, I'm new to Accelerate but like it so far from whatever i've seen! Had a couple of queries, if you could please help out that'd be awesome! 🚀  

Let's say if I do a gather on all predictions across GPUs:
```
all_preds = accelerator.gather(predictions).detach().cpu()
```
a) Does gather() have any ordering guarantees on in what order are the concatenated tensors going to be in? For example: Are they going to be concatenated in the order of GPU rank numbers or how are they concatenated in the final all_preds array? 

b) Also, was wondering say if I'm calling gather() then like in the original example posted at the beginning of this issue where we are trying to compute eval metrics after gathering all predictions (https://github.com/huggingface/accelerate/issues/226#issue-1098256277 by @jbschiratti) - then in atleast this example's case specifically wouldn't it be better to only call gather and evaluate metrics in the main_process only using `accelerator.is_main_process` or what is the advantage against doing this and allowing it to call on all processes? Can calling on main_process only lead to increased runtimes/training or eval times in general or there shouldn't be any sizeable runtime impact I'd presume? 

c) Maybe i'm not able to think of and reason out currently due to possibly my limited current understanding of Accelerate, but similarly in which case why would someone even want to call gather on all processes and not just the rank0 main process? Is there a valid use-case in support of not enforcing gather call to only run on main_process? 

Would really appreciate any help! :)

**Comment by sgugger** [comment]
Please use the [forums](https://discuss.huggingface.co/) to ask questions like those, as it's where it's going to benefit the community the most :-)

For quick answers:
a) The tensors are going to be in the exact same order as the dataset as long as you did not use `shuffle=True` or any sample, in which case it will be in the exact same order as the `dataloader` iterated through the dataset.
b-c) `gather` needs to be called on all processes as it actually does an `all_gather` behind the scenes. If you want to only evaluate on the main process, do not send the validation dataloader to the `accelerator.prepare` method and only run your evaluation loop on process 0.

**Comment by ashutoshsaboo** [comment]
Woah, that was insanely quick! ⚡  Got it, thanks - but I think you inferred my question for b-c in a diff context, let me clarify the same. So i'm ok with evaluating and comparing metrics on all GPUs since it doesn't matter (even though it's redundant work, since all gpus after gathering should give the same eval metric results in ideal cases) - which is why I was thinking why do redundant work on all GPUs in the first place. But if I have to say do wandb.log() and log results to an experiment tracker - in that case I might have to ensure that it gets called only on the main process isn't it, otherwise duplicate values might be logged from each GPU process? Or am I missing any caveat there, if you could clarify the same? @sgugger

**Comment by sgugger** [comment]
Logging should be inside a test for the main process only, yes. We will soon add utilities to help with that as well!

**Comment by lorenmt** [comment]
> Yes, especially since you could get race conditions with various processes trying to write to the same file, and a big slowdown if you try to prepare the data num_processes times.

Hello, to follow up this question. If the dataset has also been downloaded. Is it still a good practice to create datasets only on the main process? What about data loader and loading pre-trained networks? I am still confused about which part of the code should be employed only on the main process and which shouldn't?

**Comment by muellerzr** [comment]
@lorenmt can you try reading these docs here to see if they answer your questions? And if not let us know where you are still confused!

https://huggingface.co/docs/accelerate/concept_guides/deferring_execution

**Comment by lorenmt** [comment]
@muellerzr Thanks for pointing me to this page. 

I have fully read this tutorial and still have some questions.

My experiment shows that I only need to load pre-trained weights once on the main process, and it gives the same performance compared to load pre-trained weights on all processes. This makes sense because we are all updating the same model after all, and it points to the same parameters? And therefore I am wondering is it true that we only need to run once on the main process for the data loader, datasets, and optimiser as well, and then after we call the accelerator.prepare, then they will be distributed to other gpus?
